Update variables and read additional fields

michtam opened the pull request.

Commit:
Add battery level, battery volatage, light level

Commit:
Update variables

Commit:
Add battery level, battery volatage, light level and variables - debug version

bartowl commented:
Hi. I liked your work a lot! Personaly I have implemented iNode for tasmota, but it never got to a stage that it was ready for public, so it stayed private. I have checked your code, and you are missing the part, where battery_level is calculated which results in errors during compilation.
As I implemented support for iNode I did some heavy parsing too, hope this will help you. I used a bit different sensor - altough it was also an optical energy meter. There is a sample frame and the explanation. This is raw dump, so the marked section starts with manufacturer id, which is not even full 2 bytes but it also contains 2 extra flags, that might get changed in low battery condition or when memory was not read for 24h... This is what i have found in the bits and pieces of documentation available in net. This worked for me, maybe it will help you as well.
Ps. your debug byte 0 is my byte 2 below, as 0-1 is the manufacturer id. The rest should be 
```
// sample packet
// 04 3E 21 02 01 00 00 7D FA 43 18 F0 D0 15 02 01 06 0E FF [ 90 82 00 00 59 D3 02 00 DC 05 A7 00 60 ] 02 0A 08 BA
// 0: 90 82 - uint16 (0x8290) 82==iNode EM, & 0x0004 - low batt, & 0x0002 - 24h passed since last memory read
// 2: 00 00 - uint16 actual power in pulses/minute (val/base*60 in kwh or m3/h) - calculated every minute
// 4: 59 D3 02 00 - uint32 energy total since boot in pulses
// 8: DC 05 - uint16
//                    & 0x3fff - meter constant in pulses per unit (1500 here)
//                    & 0xc000 - unit (00: kw/kwh, 10: m3, other: pulses)
// 10: A7 - battery voltage (A): (10-2)*10% = 80% (detailed 0-F is 0% to 100%) volts=(((Vbat*1.2V)/10)+1.8V) + current sensor value (7) (0-15) 
// 11: 00 60 - uint16 (0x6000) - & 0xf000 - dow 0-sun..7-sat, & 0x0fff - energy in unit for last week day (0)
// 2021-10-29 00:25:40 MAC: D0F01843FA7D POWER: 0.000 kW TOTAL: 123.451 kWh PPM: 0 RSSI: -70
```
Currently I am implementing magnetic meter for gas sensor, where units are m3 and l/min, once ready i might post it here as well...
Implementation the way you did it is so smooth with esphome, not to compare with that over 500 lines of code for tasmota :)

michtam commented:
@bartowl I'm glad to hear that!

> (...) and you are missing the part, where battery_level is calculated which results in errors during compilation.

Yup, there is a big chance I've missed something ;)

> once ready i might post it here as well...

Sure, why not. Maybe it'll be useful for someone else in the future.

I have to say, I abandoned this solution and replaced it with something else - just bluetooth usb dongle and simple python script - because esphome was too slow. By _too slow_ I mean collecting data is too rare. I couldn't make it more frequently unfortunately.

However, maybe I'll give it a try and make some improvements.

bartowl commented:
In meanwhile I revised the code and changed the battery voltage calculation to match the real voltage (at least on my unit). I have connected a lab power supply instead of the battery and went from 1,8V to 3,1V with 0,1V deltas and the returned value is rather battery voltage, as it maps nearly linear to voltage change. Finally I have used own formula to calculate it which gives bang on results. For battery level - with such small discharge current the StateOfCharge of a lithium cell is way off from linear, but I just converted the returned voltage linear to 0-100% knowing that it will stay around 80-90% really long, and then drop fast down. reported 50% is rather 20% SoC but I did not want to go even deeper, as the curve depends heavily in actual discharge current which is under 1mA and hard for me to measure. I might log the voltage over its entire lifespan and after that adjust the algorithm, but  for me it is enough to know when battery is low :)

Regarding update frequency - I tuned a bit the esp32_ble_tracker - you had a duration of 1s which was waaaay to low. For me it is not unusall to get a data packet every 2-3s, so I actually had to throttle down the update rate not to spam my database with too many data points (not using homeassistant, just mqtt).

I also needed to change the manufacturer_id to match my iNodeEM, but if your value works with yours, then just use your value.

One final update is a text_sensor to report device presence, since expire_after is handled by homeassistant which I am not using, and I needed to have a way to tell when the iNode turned off (low battery? does not last long to he honest).

I got also a weird behavior (what I initially thought) from the template sensors, as even with iNode turned off they reported last value every minute. It turned out to be the default behavior of templates, and in order to disable it i used a simple lambda not to return anything. Once triggered via publish_state they work correctly, and using the throttle filter returns not more publishes than one per minute. there is also throttle_average which publishes exactly once per minute regardless if it receives data or not, averaging received samples. You may want to have a look there too. I used this as part of the presence detection - if nothing received over 5min, i assume sensor is dead. Those throttles interfered with the built-in publish of templates that happened by default every minute. I am not sure if instead of the lambda it would be sufficient to use update_interval: 0, but this one works and i will not dig probably more into this...

Here is the full code, you might find it useful at some points:
```yaml
esp32_ble_tracker:
  scan_parameters:
#    duration: 1s
    active: false
  on_ble_manufacturer_data_advertise:
    - mac_address: ${inode_mac}
      manufacturer_id: "8290" # 8290 iNodeEM
      then:
        - lambda: |-
            ESP_LOGD("ble_adv", " 0   - %02X", (x[0] ) );
            ESP_LOGD("ble_adv", " 1   - %02X", (x[1] ) );
            ESP_LOGD("ble_adv", " 2   - %02X", (x[2] ) );
            ESP_LOGD("ble_adv", " 3   - %02X", (x[3] ) );
            ESP_LOGD("ble_adv", " 4   - %02X", (x[4] ) );
            ESP_LOGD("ble_adv", " 5   - %02X", (x[5] ) );
            ESP_LOGD("ble_adv", " 6   - %02X", (x[6] ) );
            ESP_LOGD("ble_adv", " 7   - %02X", (x[7] ) );
            ESP_LOGD("ble_adv", " 8   - %02X", (x[8] ) );
            ESP_LOGD("ble_adv", " 9   - %02X", (x[9] ) );
            ESP_LOGD("ble_adv", " 10  - %02X", (x[10] ) );
            ESP_LOGD("ble_adv", " impulses hex - %02X:%02X", x[7], x[6] );
            ESP_LOGD("ble_adv", " energy hex   - %02X:%02X:%02X:%02X", x[5], x[4], x[3], x[2] );
            ESP_LOGD("ble_adv", " power hex    - %02X:%02X", x[1], x[0] );
            ESP_LOGD("ble_adv", " bat&lgt hex  - %02X", x[8] );
            uint16_t impulses_raw = (x[6])  + ((x[7] & 0x3f) << 8);
            uint16_t power_raw = (x[0]) + (x[1] << 8);
            uint32_t energy_raw = (x[2]) + (x[3] << 8) + (x[4] << 16) + (x[5] << 24);
            int battery_raw = (x[8] >> 4);
            int light_raw = (x[8] & 0x0F );
            ESP_LOGD("ble_adv", " impulses_raw - %i", impulses_raw );
            ESP_LOGD("ble_adv", " power_raw    - %i", power_raw );
            ESP_LOGD("ble_adv", " energy_raw    - %i", energy_raw );
            ESP_LOGD("ble_adv", " battery_raw    - %i", battery_raw );
            ESP_LOGD("ble_adv", " light_raw    - %i", light_raw );
            float power = float(power_raw)/float(impulses_raw) * 1000;
            float energy = float(energy_raw)/float(impulses_raw);
            float battery_voltage = (float(battery_raw)/10) + 1.8;
            int battery_level = min(((battery_raw - 2) * 10),100);
            int sensor_level = int((float(light_raw) * 100) / 15.0);
            ESP_LOGD("ble_adv", " power    - %.2f l/min", power );
            ESP_LOGD("ble_adv", " energy    - %.2f m3", energy );
            ESP_LOGD("ble_adv", " battery_level    - %i %%", battery_level );
            ESP_LOGD("ble_adv", " battery_voltage    - %.2f V", battery_voltage );
            ESP_LOGD("ble_adv", " sensor_level    - %i %%", sensor_level );
            id(inode_power).publish_state(power);
            id(inode_energy).publish_state(energy);
            id(inode_impulses).publish_state(impulses_raw);
            id(inode_battery_voltage).publish_state(battery_voltage);
            id(inode_battery_level).publish_state(battery_level);
            id(inode_sensor_level).publish_state(sensor_level);

text_sensor:
  - platform: template
    name: "inode presence"
    id: inode_presence
    lambda: return {}; # no periodic updates from template

sensor:
  - platform: uptime
    name: "Uptime"
    icon: mdi:clock-outline
    update_interval: 60s
    
  - platform: wifi_signal
    name: "WiFi Signal"
    update_interval: 60s

  ######################################### INODE
  - platform: ble_rssi
    mac_address: ${inode_mac}
    name: "inode RSSI"
    filters:
      - throttle_average: 60s
      - lambda: |-
          if (isnan(x)) {
            id(inode_presence).publish_state("absent");
            return {}; // do not publish new RSSI, just update presence
          } else  {
            id(inode_presence).publish_state("present");
            return x;
          }

  - platform: template
    name: "inode Power"
    id: inode_power
    unit_of_measurement: "l/min"
    accuracy_decimals: 3
    icon: "mdi:flash"
    lambda: return {}; # no periodic updates from template
    filters:
      - throttle: 60s

  - platform: template
    name: "inode Total Energy"
    id: inode_energy
    unit_of_measurement: 'm3'
    accuracy_decimals: 2
    icon: "mdi:flash"
    lambda: return {}; # no periodic updates from template
    filters:
      - throttle: 60s

  - platform: template
    name: "inode Impulses"
    id: inode_impulses
    unit_of_measurement: 'imp/m3'
    accuracy_decimals: 0
    lambda: return {}; # no periodic updates from template
    filters:
      - throttle: 3600s # once per hour

  - platform: template
    name: "inode Battery-level"
    id: inode_battery_level
    unit_of_measurement: '%'
    accuracy_decimals: 0
    lambda: return {}; # no periodic updates from template
    filters:
      - throttle: 600s # every 10 min

  - platform: template
    name: "inode Battery-voltage"
    id: inode_battery_voltage
    unit_of_measurement: 'V'
    accuracy_decimals: 3
    lambda: return {}; # no periodic updates from template
    filters:
      - throttle: 600s # every 10 min

  - platform: template
    name: "inode Sensor level"
    id: inode_sensor_level
    unit_of_measurement: '%'
    accuracy_decimals: 0
    lambda: return {}; # no periodic updates from template
    filters:
      - throttle: 60s

```

bartowl commented:
I also get very good response from tuned scan_parameters as described here: https://esphome.io/components/sensor/xiaomi_ble.html?highlight=xiaomi#improving-reception-performance

it might be also worth disabling logger component completely, as it pushes a lot of mqtt messages, and wifi and bluetooth share the same bandwidth, so where esp32 sends something over wifi it does not hear anything over ble. Or at least limit verbosity to informational.

Those things should increase the "speed" of detecting devices...

michtam commented:
First of all, sorry @bartowl for that long response. It's not on my top priority list.

> In meanwhile I revised the code and changed the battery voltage calculation to match the real voltage (at least on my unit). I have connected a lab power supply instead of the battery and went from 1,8V to 3,1V with 0,1V deltas and the returned value is rather battery voltage, as it maps nearly linear to voltage change. Finally I have used own formula to calculate it which gives bang on results. For battery level - with such small discharge current the StateOfCharge of a lithium cell is way off from linear, but I just converted the returned voltage linear to 0-100% knowing that it will stay around 80-90% really long, and then drop fast down. reported 50% is rather 20% SoC but I did not want to go even deeper, as the curve depends heavily in actual discharge current which is under 1mA and hard for me to measure. I might log the voltage over its entire lifespan and after that adjust the algorithm, but for me it is enough to know when battery is low :)

Whoa, I'm impressed that someone put so much effort on this :D

> I used this as part of the presence detection - if nothing received over 5min, i assume sensor is dead.

In my case it may happen I don't receive anything for 1,5h - even using usb BT dongle and scan everything around.

> Here is the full code, you might find it useful at some points:

Cool. It's time to refresh code and test it ;)

bartowl commented:
No problem, I adjusted it for my needs and thanks to your code got already a nice heads up, so why not to share. Increasing throttle_average to 1,5-2h should also work if you don't mind averaging of the rssi over such large timeframe. Thanks!

oomichalekoo commented:
power ,total and impulses dont work i cant compile becouse mu manufactur id is 0x45 somebody can help me ?

oomichalekoo commented:
> esphome:
  name: esp32_inode
  platform: ESP32
  board: nodemcu-32s

wifi:
  ssid: ASUS_2.4
  password: !secret password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Esp32 BT Fallback Hotspot"
    password: "JdytF7dpLgSs"

captive_portal:

# Enable logging
#logger:

# Enable Home Assistant API - integracja ESPHome
#api:
#  password: !secret password

# Odkomentuj jeśli chcesz mieć stronkę
#web_server:
#  port: 80

mqtt:
  broker: 192.168.1.38
  port: 1883
  discovery: True
  
ota:
  password: !secret ota_password



esp32_ble_tracker:
  scan_parameters:
    duration: 1s
    active: false
  on_ble_manufacturer_data_advertise:
    - mac_address: "D0:F0:18:78:0B:C1"
      manufacturer_id: "8290"
      then:
        - lambda: |-
            ESP_LOGD("ble_adv", " 0   - %02X", (x[0] ) );
            ESP_LOGD("ble_adv", " 1   - %02X", (x[1] ) );
            ESP_LOGD("ble_adv", " 2   - %02X", (x[2] ) );
            ESP_LOGD("ble_adv", " 3   - %02X", (x[3] ) );
            ESP_LOGD("ble_adv", " 4   - %02X", (x[4] ) );
            ESP_LOGD("ble_adv", " 5   - %02X", (x[5] ) );
            ESP_LOGD("ble_adv", " 6   - %02X", (x[6] ) );
            ESP_LOGD("ble_adv", " 7   - %02X", (x[7] ) );
            ESP_LOGD("ble_adv", " 8   - %02X", (x[8] ) );
            ESP_LOGD("ble_adv", " 9   - %02X", (x[9] ) );
            ESP_LOGD("ble_adv", " 10  - %02X", (x[10] ) );
            ESP_LOGD("ble_adv", " impulses hex - %02X:%02X", x[7], x[6] );
            ESP_LOGD("ble_adv", " energy hex   - %02X:%02X:%02X:%02X", x[5], x[4], x[3], x[2] );
            ESP_LOGD("ble_adv", " power hex    - %02X:%02X", x[1], x[0] );
            ESP_LOGD("ble_adv", " bat&lgt hex  - %02X", x[8] );
            uint16_t impulses_raw = (x[6])  + (x[7] << 8);
            uint16_t power_raw = (x[0]) + (x[1] << 8);
            uint32_t energy_raw = (x[2]) + (x[3] << 8) + (x[4] << 16) + (x[5] << 24);
            ESP_LOGD("ble_adv", " impulses_raw - %i", impulses_raw );
            ESP_LOGD("ble_adv", " power_raw    - %i", power_raw );
            ESP_LOGD("ble_adv", " energy_raw    - %i", energy_raw );
            float power = float(power_raw)/float(impulses_raw) * 60.0 * 2500;
            float energy = float(energy_raw)/float(impulses_raw);
            id(inode_power).publish_state(power);
            id(inode_energy).publish_state(energy);
            id(inode_impulses).publish_state(impulses_raw);
            
sensor:
  - platform: ble_rssi
    mac_address: "D0:F0:18:78:0B:C1"
    name: "inode"

  - platform: uptime
    id: socket_uptime
    name: "Uptime"
    icon: mdi:clock-outline
    update_interval: 60s
    
  - platform: wifi_signal
    id: socket_wifi_signal
    name: "WiFi Signal"
    update_interval: 30s
    
  - platform: template
    name: "inode_power"
    id: inode_power
    unit_of_measurement: "W"
    accuracy_decimals: 3
    icon: "mdi:flash"
    
  - platform: template
    name: "Total Energy"
    id: inode_energy
    unit_of_measurement: 'kWh'
    accuracy_decimals: 3
    icon: "mdi:flash"
    
  - platform: template
    name: "Impulses"
    id: inode_impulses
    unit_of_measurement: 'imp/kWh'
    accuracy_decimals: 0
    
  - platform: total_daily_energy
    name: "Total Daily Energy"
    power_id: inode_power
    unit_of_measurement: kW
    filters:
    - multiply: 0.001
